### PR TITLE
Support using Kaniko to build images.

### PR DESCRIPTION
**Pull request by xingao267**

### CODE REVIEW

File: contrib/dockerfile_build.bzl [hunk 1]
```diff
@@ -15,9 +15,10 @@
 """Repository rule for building a Docker image from a Dockerfile and saving it
    into a tar file.
 
-This rule uses the docker tool installed in the system to build the image and
-then save it as a tar file. The saved tar file is available for other rules to
-use (e.g. container_image, container_load, container_test)
+This rule uses the docker tool installed in the system to build the image or run
```
Comment (nlopezgi) [inline contrib/dockerfile_build.bzl]: This rule uses either the docker tool installed in the system or the Kaniko executor container to build the image and then save it as a tar file.

File: contrib/dockerfile_build.bzl [hunk 2]
```diff
@@ -59,63 +70,165 @@ def _impl(repository_ctx):
...
         "build_args": attr.string_dict(
             doc = "A map of args to pass to the --build-arg option in the " +
-                  "`docker build` command.",
+                  "docker or kaniko build command.",
         ),
         "docker_path": attr.string(
-            doc = "The full path to the docker binary. If not specified, it " +
-                  "will be searched for in the path. If not available, " +
-                  "the rule build will fail.",
+            doc = "The full path to the docker binary.",
```
Comment (nlopezgi) [inline contrib/dockerfile_build.bzl]: Are we no longer trying to find it on path?
Make mandatory if so, if not leave comment as was before?

File: contrib/dockerfile_build.bzl [hunk 3]
```diff
@@ -59,63 +70,165 @@ def _impl(repository_ctx):
...
         ),
         "dockerfile": attr.label(
             allow_single_file = True,
             mandatory = True,
             doc = "The label for the Dockerfile to build the image from.",
         ),
+        "driver": attr.string(
```
Comment (nlopezgi) [inline contrib/dockerfile_build.bzl]: What's the default?
I see default below, make mandatory then

File: contrib/dockerfile_build.bzl [hunk 4]
```diff
@@ -59,63 +70,165 @@ def _impl(repository_ctx):
...
+    if kaniko_digest and not kaniko_digest.startswith("sha256:"):
+        fail("kanoko_digest must start with 'sha256:'.")
+
+    kaniko_image_path = _KANIKO_IMAGE_PATH
+    if kaniko_digest:
+        kaniko_image_path = _KANIKO_IMAGE_PATH + "@" + kaniko_digest
+    if kaniko_tag:
+        kaniko_image_path = _KANIKO_IMAGE_PATH + ":" + kaniko_tag
+
+    _dockerfile_image(
```
Comment (nlopezgi) [inline contrib/dockerfile_build.bzl]: You need to pass the docker_path

File: contrib/dockerfile_build.bzl [hunk 5]
```diff
@@ -59,63 +70,165 @@ def _impl(repository_ctx):
...
         "dockerfile": attr.label(
             allow_single_file = True,
             mandatory = True,
             doc = "The label for the Dockerfile to build the image from.",
         ),
+        "driver": attr.string(
+            doc = "The image building tool to use. Currently, `docker` and " +
+                  "`kaniko` are supported.",
+        ),
+        "kaniko_image_path": attr.string(
```
Comment (nlopezgi) [inline contrib/dockerfile_build.bzl]: make mandatory

File: contrib/kaniko_run_and_extract.sh.tpl [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+set -ex
+
+# This is a generated file that runs Kaniko inside a docker container, waits for
+# it to finish running and copies the generated image tarball out of it.
+
+# Copy the build context into the container.
+data=$(%{docker_path} create -v %{kaniko_workspace} %{image_path})
+%{docker_path} cp %{build_context_dir}/. ${data}:%{kaniko_workspace}
+
+# Pass an empty entrypoint to override any set by default in the container.
```
Comment (nlopezgi) [inline contrib/kaniko_run_and_extract.sh.tpl]: I dont see the entrypoint below, update comment?

File: tests/docker/BUILD [hunk 7]
```diff
@@ -358,9 +358,13 @@ docker_push_all(
     bundle = ":three_images_bundle",
 )
 
-container_test(
-    name = "basic_dockerfile_image",
-    configs = ["//tests/docker/configs:basic_dockerfile.yaml"],
+[container_test(
```
Comment (nlopezgi) [inline tests/docker/BUILD]: Lets move these tests (if possible) to GCB (sync with Alex that is also doing this for some other tests)

File: tests/docker/configs/dockerfile_image.yaml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    expectedContents: ['Some data\.\.\.\s']
+  - name: 'usr_txt_file_contains'
+    path: '/usr/More_data.txt'
+    expectedContents: ['Additional data']
+  - name: 'copy_txt_file_contains'
+    path: '/file_to_copy.txt'
+    expectedContents: ['Another file to copy']
+
+# TODO(xingao): Enable these tests once "diff_id" key not found in image error
+# is fixed in container_bundle.
```
Comment (nlopezgi) [inline tests/docker/configs/dockerfile_image.yaml]: Is there some issue or something to track this?

File: contrib/dockerfile_build.bzl (hunk 1 above)
Comment (xingao267) [reply contrib/dockerfile_build.bzl]: Done.

File: contrib/dockerfile_build.bzl (hunk 2 above)
Comment (xingao267) [reply contrib/dockerfile_build.bzl]: Fixed. I added more detailed comments below in the macro docs.

File: contrib/dockerfile_build.bzl (hunk 3 above)
Comment (xingao267) [reply contrib/dockerfile_build.bzl]: Done.

File: contrib/dockerfile_build.bzl (hunk 5 above)
Comment (xingao267) [reply contrib/dockerfile_build.bzl]: Done.

File: contrib/dockerfile_build.bzl (hunk 4 above)
Comment (xingao267) [reply contrib/dockerfile_build.bzl]: oops...fixed.

File: contrib/kaniko_run_and_extract.sh.tpl (hunk 6 above)
Comment (xingao267) [reply contrib/kaniko_run_and_extract.sh.tpl]: Done.

File: tests/docker/configs/dockerfile_image.yaml (hunk 8 above)
Comment (xingao267) [reply tests/docker/configs/dockerfile_image.yaml]: Created https://github.com/bazelbuild/rules_docker/issues/903

File: tests/docker/BUILD (hunk 7 above)
Comment (xingao267) [reply tests/docker/BUILD]: Discussed offline. These tests are running on Buildkite + RBE and runs pretty fast.

**Pull request merged**